pybind11/setup_helpers.py

```diff
@@ -0,0 +1,158 @@
+This module provides a way to check to see if flag is available,
+has_flag (built-in to distutils.CCompiler in Python 3.6+), and
+a cpp_flag function, which will compute the highest available
+flag (or if a flag is supported).
+
+LICENSE:
+
+Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>, All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import os
+import sys
+import tempfile
+
+import distutils.errors
+from distutils.command.build_ext import build_ext
+
+
+# It is recommended to use PEP 518 builds if using this module. However, this
+# file explicitly supports being copied into a user's project directory
+# standalone, and pulling pybind11 with the deprecated setup_requires feature.
+
+
+class DelayedPybindInclude(object):
+    """
+    Helper class to determine the pybind11 include path The purpose of this
+    class is to postpone importing pybind11 until it is actually installed, so
+    that the ``get_include()`` method can be invoked if pybind11 is loaded via
+    setup_requires.
+    """
+
+    def __str__(self):
+        import pybind11
+
+        return pybind11.get_include()
+
+
+# cf http://bugs.python.org/issue26689
+def has_flag(compiler, flagname):
+    """
+    Return a boolean indicating whether a flag name is supported on the
+    specified compiler.
+    """
+
+    with tempfile.NamedTemporaryFile("w", suffix=".cpp", delete=False) as f:
+        f.write("int main (int argc, char **argv) { return 0; }")
+        fname = f.name
+    try:
+        compiler.compile([fname], extra_postargs=[flagname])
```

Comment:
There's an issue where this command will create temporary files in the current directory instead of the system $TMP. Here's the fix:

```python
            # distutils/ccompiler.py, unixcompiler.py, etc.
            # compiler.compile generates output file at
            # os.path.join(output_dir, fname[1:]) - which drops leading /,
            # so we use output_dir == '/' to put it back on
            # not sure what Windows does so leave it alone
            outdir = sep if sys.platform != 'win32' else None
            compiler.compile([fname], extra_postargs=[flagname], output_dir=outdir)
```

Reply:
(Fixed but will try to test on Windows)

Reply:
I've implemented the fix in a more general way now that should be fine on Windows too (builds in a temp directory)